src/families.h

```diff
@@ -159,128 +126,149 @@ class Binomial : public Family {
+    gradient[0] = 1.0 - y[i] - 1.0/(1.0 + std::exp(prediction[0]));
   }
 
-  double NullDeviance(const bool fit_intercept) {
-    double pred = fit_intercept ? Link(Mean(y)) : 0.0;
+  double NullDeviance(const std::vector<double>& y,
+                      const bool                 fit_intercept,
+                      const unsigned             n_classes) const noexcept {
+    double y_mu = fit_intercept ? Link(Mean(y)) : 0.0;
+    std::vector<double> prediction(1, y_mu);
 
     double loss = 0.0;
-    for (unsigned i = 0; i < n_samples; ++i)
-      loss += Loss(pred, i);
+    for (unsigned i = 0; i < y.size(); ++i)
+      loss += Loss(prediction, y, i);
 
     return 2.0 * loss;
   }
-};
-
-class Multinomial : public Family {
-public:
-  Multinomial(const std::vector<double>& y,
-              const unsigned n_samples,
-              const unsigned n_classes)
-              : Family(y, n_samples, n_classes, 0.5) {}
 
-  void PreprocessResponse() {
-    std::vector<double> y_mu(n_classes);
-    std::vector<double> y_var(n_classes);
+  Eigen::MatrixXd
+  LambdaResponse(const std::vector<double>& y,
+                 std::vector<double>&       y_mat_scale) const noexcept {
 
-    for (unsigned i = 0; i < n_samples; ++i) {
-      unsigned y_class = std::round(y[i]);
-      y_mat(i, y_class) = 1.0;
-    }
+    Eigen::MatrixXd y_mat(y.size(), 1);
 
-    for (unsigned j = 0; j < n_classes; ++j)
-      y_mu[j] = y_mat.col(j).mean();
+    auto y_mu = Mean(y);
+    auto y_sd = StandardDeviation(y, y_mu);
 
-    for (unsigned i = 0; i < n_classes; ++i) {
-      for (unsigned j = 0; j < n_samples; ++j)
-        y_var[i] += std::pow(y_mat(j, i) - y_mu[i], 2)/n_samples;
+    for (unsigned i = 0; i < y.size(); ++i)
+      y_mat(i, 0) = (y[i] - y_mu)/y_sd;
 
-      y_mat_scale[i] = y_var[i] == 0.0 ? 1.0 : std::sqrt(y_var[i]);
-    }
+    y_mat_scale[0] = y_sd;
 
-    for (unsigned i = 0; i < n_classes; ++i) {
-      for (unsigned j = 0; j < n_samples; ++j) {
-        y_mat(j, i) -= y_mu[i];
-        y_mat(j, i) /= y_mat_scale[i];
-      }
-    }
+    return y_mat;
   }
+};
 
-  double Link(const unsigned i) {
-    return std::log(y[i] / (1.0 - y[i]));
-  }
+class Multinomial : public Family {
+public:
+  Multinomial() : Family(0.25) {}
 
-  double Loss(const double prediction, const unsigned i) {
-    // Not reasonable for multinomial model
-    return 0.0;
+  void Preprocess(std::vector<double>& y,
+                  std::vector<double>& y_center,
+                  std::vector<double>& y_scale) const noexcept {
+    // no preprocessing
   }
 
-  double Loss(const std::vector<double>& prediction, const unsigned i) {
-
-    unsigned y_class = std::round(y[i]);
-
-    return LogSumExp(prediction) - prediction[y_class];
+  double Loss(const std::vector<double>& prediction,
+              const std::vector<double>& y,
+              const unsigned             i) const noexcept {
+    auto c = static_cast<unsigned>(y[i] + 0.5);
```

Comment:
Can you comment this function a bit more? I'm having trouble following what happens: why do we add 0.5 to the label and then round to an integer? Shouldn't `y` be an integer to start with?

Reply:
Sure. I did this to avoid needing to template on double/integer. I don't think there is any real overhead with the static casts, although there is certainly a memory advantage with storing integers rather than doubles. Hm. Perhaps the response should go into the family class anyway. I'm not sure.

The addition of 0.5 is there because casting `double` to `int` is essentially truncation and because 1 might very well be represented as 0.99999999999 (or something like it), we might end up casting "1" to 0.0 if we did not add 0.5.

Reply:
Hmmmm... ok, that makes sense, but is probably worth commenting (if nothing else)